Add delete entry function

jinfeihan57 opened the pull request:
@kuba-- 
I have implemented the entry deletion function. The idea is to obtain the offset and data length of every file data in the ZIP file first, then move the data behind the deleted entry forward to overwrite the deleted data, and in the  end truncate the invalid data at the end of the ZIP file. The case of deleting one folder (multiple files) at a time is considered in the implementation.

kuba-- commented:
Wow, thanks! 
I'll take a look. In a meantime, can you check why CI is failing.

jinfeihan57 commented:
@kuba-- 
Sorry, there were some accidents before submitting the code, I've fixed it.

kuba-- reviewed (commented):
I'll continue, slowly review the PR.

jinfeihan57 commented on CMakeLists.txt:
```diff
@@ -45,7 +45,7 @@ if (MSVC)
         "${CMAKE_C_COMPILER_ID}" STREQUAL "Clang" OR
         "${CMAKE_C_COMPILER_ID}" STREQUAL "AppleClang")
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c99 -Wall -Wextra -Werror -pedantic")
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu99 -Wall -Wextra -Werror -pedantic")
```
-std=gnu99 is required by the function ftruncate().

jinfeihan57 commented on src/zip.c:
```diff
@@ -1023,6 +1081,19 @@ int zip_extract_stream(const char *stream, size_t size, const char *dir,
   return status;
 }
 
+typedef enum {
+  KEEP = 0,
+  DELETE = 1,
+  MOVE = 2,
+} modify_type;
+
+struct entry_mark {
+  int file_index;
+  modify_type type;
+  mz_uint64 m_local_header_ofs;
+  mz_uint64 lf_length;
+};
+
```
When I want to  delete one folder (multiple files) at a time,  I need to  record the information about all files to be deleted and then delete them one by one. The struct entry_mark is for this perpose. An entry_mark element keeps the information of a file, An array of entry_mark element keeps the  information of all files in .ZIP file. The index of entry_mark element in the array shows the occurrence order of a file in  central directory structure, the variables in the structure are described as follows:

`struct entry_mark {`
  `   int file_index;     // shows the order in which local file header related with this entry  appears  `
  `   modify_type type;    //  shows the file related with this entry should be kept, deleted, or moved?`
   `  mz_uint64 m_local_header_ofs;  // keeps the offset of file data(local file header + file data + data descriptor )  relative to the start of the .ZIP file`
  `   mz_uint64 lf_length;  //  keeps the file data(local file header + file data + data descriptor ) length`
`};`

jinfeihan57 replied:
By the way, this pr supports the deletion of multiple files at a time a at the bottom level, but it has not been implemented at the interface level, I'm working on it.

kuba-- replied:
Maybe we can simplify it and expose api which let you delete entries one by one.
I'm not 100% sure, but if you take a look into zip command (delete entries parameter), then for zip file:
```
zipinfo tmp.zip 
Archive:  tmp.zip
Zip file size: 1378 bytes, number of entries: 10
drwxr-xr-x  3.0 unx        0 bx stor 20-Oct-25 02:04 a/
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:04 a/a3
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:04 a/a2
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:04 a/a1
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:04 a/c2
drwxr-xr-x  3.0 unx        0 bx stor 20-Oct-25 02:05 b/
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:05 b/b2
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:05 b/b3
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:05 b/b1
drwxr-xr-x  3.0 unx        0 bx stor 20-Oct-25 02:04 empty/

zip tmp.zip -d "a/"
deleting: a/

zip tmp.zip -d "a/*"
deleting: a/a3
deleting: a/a2
deleting: a/a1
deleting: a/c2
```

I'm not saying to implement _regexp_ like `dir/*`, but maybe for simplicity `dir/` will remove just this particular entry name.
WDYT?

jinfeihan57 replied:
The current pr supports to delete a file or a folder at a time. The process of the current PR is as follows:

1. Open .zip file, initialize a data structure `mz_zip_archive`. 
2. create an array `entry_mark_array` to keep the information(data offset, data length, if the data needs to be deleted, the mapping relationship between the entries in the central directory and the file data blocks) about all compressed files 
3. Traverse `entry_mark_array`, delete all file data and modify the data which is affected by the deleted file in the central directory in one traversal. Then delete the entrys in the central directory and truncate the file at the end.
(If the .zip file contains a dir/ and dir/* , we can't delete the entry dir/ alone now, this is a little bug, I can fix it and implement regexp like dir/* by modify several lines of code. )

If we just creat a simple api which let people delete entries one by one. I think the api process can be as follows(there may be a simpler way to do it):

1. Open .zip file, initialize a data structure `mz_zip_archive`.
2. Get the information that needed for deleting a file. The information includes the length and the offset relative to the file start position of the data block to be deleated . then modify the data which is affected by the deleted file in the central directory.
 3. move the data block behind the deleated data block forward to overlay the deleated data block,  and truncate the file at the end.

   In step 2, we can  call the function `zip_entry_open(zip, deleted_file_name)` to get some information that needed for deleting a file. We can only get `entry index`, `m_central_dir_ofs` and `m_local_header_ofs`. to get data length we have other work to do. Since the  [.ZIP file format spec](https://pkware.cachefly.net/webdocs/APPNOTE/APPNOTE-6.2.0.txt) says `"The entries in the central directory may not necessarily be in the same order that files appear in the .ZIP file"`, to delete data correctly from a .zip file, we still need to create a data structure to keep the index, offset and length of all file data blocks(local file header + file data + data descriptor).what's more, we need to record the association between the entries in the central directory and the file data blocks. 

  So in my opinion, we have limited space for simplifying the code if we just expose a simple api. On the other hand, we will waste a lot of time in initializing auxiliary data structures when deleting folders by using a simple API, especially when there are many files in the folder. I think it would be very useful to support deleting folders. But if we have a special need to implment a simple api, I can do it too.

jinfeihan57 replied:
@kuba-- 
I've updated the code and  improved the current API. Although I know you want a simple API, not a complex API with full functions. But it's hard to reduce a lot of code when simpling the current API.  So I just made a little change to implement the regexp.  Do you have any ideas for implementing a simple API?

jinfeihan57 commented:
@kuba-- 
I have finished the function to delete folders and refactored the code. You can continue to  review it.

kuba-- reviewed (commented):
I'll continue review this PR (because it's important and quite big change), so I'm sending partial review.
Thanks.

jinfeihan57 commented on src/zip.c:
```diff
@@ -197,6 +233,15 @@ struct zip_t *zip_open(const char *zipname, int level, char mode) {
       goto cleanup;
     }
     break;
+  case 'd':
+    if (zip_reader_writer_init_file(
+            &(zip->archive), zipname,
+            zip->level | MZ_ZIP_FLAG_DO_NOT_SORT_CENTRAL_DIRECTORY)) {
+      // An archive file does not exist or cannot initialize
+      // zip_archive reader
+      goto cleanup;
+    }
+    break;
```
Yeah, we can simplify the code here!  Since the value of the variable `pZip->m_central_directory_file_ofs` is reset in function `mz_zip_writer_init_from_reader()`, we just need to replace `m_central_directory_file_ofs` with `m_archive_size` in the subsequent code.

jinfeihan57 commented on src/zip.c:
```diff
@@ -209,12 +254,25 @@ struct zip_t *zip_open(const char *zipname, int level, char mode) {
   return NULL;
 }
 
+int zip_file_resize(mz_zip_archive *pZip) {
```
That's a good idea! I will revise it.

jinfeihan57 commented on src/zip.c:
```diff
@@ -1038,3 +1109,402 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
+    return MZ_FALSE;
+  }
+
+  if (delete_name_length > file_name_length) {
+    CLEANUP(delete_entry_name);
+    return MZ_FALSE;
+  }
+  if (delete_entry_name[delete_name_length - 1] == '/') {
+    if (strncmp(file_name, delete_entry_name, delete_name_length) == 0) {
+      CLEANUP(delete_entry_name);
+      return MZ_TRUE;
+    }
+    CLEANUP(delete_entry_name);
+    return MZ_FALSE;
+  }
+  if (strcmp(file_name, delete_entry_name) == 0) {
+    CLEANUP(delete_entry_name);
+    return MZ_TRUE;
+  }
+
+  CLEANUP(delete_entry_name);
+  return MZ_FALSE;
```
OK，I will revise it.

jinfeihan57 commented:
@kuba-- 
I've updated the code and improved the current API. Although I know you want a simple API, not a complex API with full functions. But it's hard to reduce a lot of code when simpling the current API. So I just made a little change to implement the regexp. Do you have any ideas for creating a simple API?

kuba-- commented:
> @kuba--
> I've updated the code and improved the current API. Although I know you want a simple API, not a complex API with full functions. But it's hard to reduce a lot of code when simpling the current API. So I just made a little change to implement the regexp. Do you have any ideas for creating a simple API?


Thanks, I'll take a look. It's kind of hectic time for me, so I have to take turns and review it in multiple steps.
Generally no worries about over simplifying stuff. Also regexp is super syntactic feature, but I can imagine it can complicate the code, so I'm not 100% sure it's worth to put the effort there.

I more worry about not-breaking things and performance.
If it works and performance is good (actually I didn't run any benchmarks, like with 10K files and remove every second), then it's OK, to merge it.
But also the code is growing, so I'd like to understand to handle future maintenance easier.

jinfeihan57 commented:
@kuba-- 
I did a simple contrast test on my machine. I create a .zip file whith a folder which contains 10k files.  Then I used our zip and Zip 3.0 in linux  separately  to delete files one by one and delete the entire folder. The result is as follows:

software |totle time(delete 10K files   one by one)| averge time(delete 10K files   one by one ) | time(delete folder)
-- | -- | -- | --
kuba--/zip |117.404172 seconds| 11.7404172 ms | 23.343 ms
Zip 3.0 |80.895seconds| 8.0895 ms |31 ms

As shown in the table. When we delete a folder which contains a lot of files, deleting files one by one will waste a lot of time. 
So support deleting folders is very valuable.
(My test case can be seen [here](https://github.com/jinfeihan57/zip/blob/benchmark/test/test.c).  )

Another thing：I have improved the code several times, the commit log is dirty due to the  wrong use of  `git rebase`💔, please clean the git log before merging it.

kuba-- commented on test/test.c:
```diff
@@ -547,6 +547,180 @@ static void test_open_stream(void) {
 #endif
 }
 
+static int creat_zip_file(const char *filename) {
```
`s/creat/create/`

jinfeihan57 replied:
OK, I will fix it.

kuba-- reviewed (changes requested):
I know _delete dir_ feature is useful, but I still wonder about it. First of all it complicates the logic a little bit. The second _directory_ is kind of artificial _entry_.
In zip archive we talk about particular entries, so I thought more about expose API, which let us delete multiple _any_ entries. For instance something super simple:
```c
z = zip_open("file.zip", 0, "d");
zip_delete_entries(z, char *entries[], int n);
zip_close(z);
```

So you can't open any entry (and generally work with any entry) when you opened archive in "d" mode.
you can just call `zip_delete_entries`, which basically will mark entries (which should be deleted, which moved and which kept).
All the stuff will be done, when you call `zip_close` (and you're in "d" mode).

jinfeihan57 commented on src/zip.c:
```diff
@@ -1038,3 +1074,397 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
+    } else {
+      entry_mark_array[i].type = KEEP;
+    }
+    CLEANUP(zip->entry.name);
```
When we continuously delete files as follows, this place will cause a memory leak:
  `assert(1 == zip_entry_delete(zip, "test/test-1.txt"));`
  `assert(1 == zip_entry_delete(zip, "test\\test-3.txt"));`
  `assert(1 == zip_entry_delete(zip, "test/empty/"));`
  `assert(1 == zip_entry_delete(zip, "empty/"));`

kuba-- replied:
ohh, right because so far all functions `zip_entry_...` required `open/close`.
That's why I was thinking about just `zip_entries_delete`

jinfeihan57 commented on src/zip.c:
```diff
@@ -1038,3 +1074,397 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
+  if (delete_name_length - 1 > file_name_length) {
+    goto cleanup;
+  }
+  if ((delete_entry_name[delete_name_length - 2] == '/') &&
```
This is a problem, I will fix it.

jinfeihan57 commented:
If the entries in the central directory are in the same order that files appear in the .ZIP file, everything will be simple. But they may be not, so we need extra data structures to hold all entry information to make sure we can delete the files correctly. This complicates the code logic.

If we were to implement a simple API, the code would be complex too. If we add the support to delete multiple files at a time based on the simple API, the code will be just a little more complex but the performance will be improved a lot. I think the proper complexity is worth it, unless we need to guarantee the speed of deleting individual files.

I think you want 'zip_delete_entries(z, char *entries[], int n)' rather than delete one folder at a time, This requires only a few changes to the existing code. Am I understand right?

kuba-- commented:
> I think you want 'zip_delete_entries(z, char *entries[], int n)' rather than delete one folder at a time, This requires only a few changes to the existing code. Am I understand right? Deleting folders is very useful, it is also implemented by other software, shall we keep it?

Yes, it will take an array of `n` entry names and mark what needs to be deleted, moved, kept.
and `zip_close` for "d" mode will do the rest.

jinfeihan57 commented:
> > I think you want 'zip_delete_entries(z, char *entries[], int n)' rather than delete one folder at a time, This requires only a few changes to the existing code. Am I understand right? Deleting folders is very useful, it is also implemented by other software, shall we keep it?
> 
> Yes, it will take an array of `n` entry names and mark what needs to be deleted, moved, kept.
> and `zip_close` for "d" mode will do the rest.

The second parameter char *entries[] is a little complicated when delete A entries not several entries.
what about a new API (zip_delete_entrie_list(z, char *entries[], int n)) for delete several entries. so we can keep API zip_delete_entries() for A file or A folder.

kuba-- commented:
> > > I think you want 'zip_delete_entries(z, char *entries[], int n)' rather than delete one folder at a time, This requires only a few changes to the existing code. Am I understand right? Deleting folders is very useful, it is also implemented by other software, shall we keep it?
> > 
> > 
> > Yes, it will take an array of `n` entry names and mark what needs to be deleted, moved, kept.
> > and `zip_close` for "d" mode will do the rest.
> 
> The second parameter char *entries[] is a little complicated when delete A entries not several entries.
> what about a new API (zip_delete_entrie_list(z, char *entries[], int n)) for delete several entries. so we can keep API zip_delete_entries() for A file or A folder.


I think, I don't fully understand what you mean, or maybe what's the difference.
I thought about following use case:
```c
z  = zip_open(file, 0, "d");

char *entries[] = {"file.txt", "a", "directory/file.1", "otherdirectory/file.3", "directory/file.2"};
zip_entries_delete(z, entries, 5);

zip_close(z);
```

jinfeihan57 commented:
> I thought about following use case:
> 
> ```c
> z  = zip_open(file, 0, "d");
> 
> char *entries[] = {"file.txt", "a", "directory/file.1", "otherdirectory/file.3", "directory/file.2"};
> zip_entries_delete(z, entries, 5);
> 
> zip_close(z);
> ```

Ok, I will change the API exposed.

jinfeihan57 commented:
@kuba--As you said, I have modified the API. Please check again.

kuba-- commented on src/zip.c:
```diff
@@ -1038,3 +1074,395 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
+  length[n - 1] = zip->archive.m_archive_size - local_header_ofs_array[n - 1];
+
+  for (int i = 0; i < n; i++) {
+    entry_mark_array[i].lf_length = length[entry_mark_array[i].file_index];
```
Can we calculate it in one loop (above)?

jinfeihan57 replied:
Most likely not. 
According to
[4.4.1.3  The entries in the central directory MAY NOT necessarily
      be in the same order that files appear in the .ZIP file.](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT)
The order of length and entry_mark_array may be different.
so : entry_mark_array[i].lf_length = length[entry_mark_array[i].file_index];
It's not the same order.

kuba-- commented on src/zip.c:
```diff
@@ -1038,3 +1074,395 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
+  return 0;
+}
+
+static int delete_central_dir_entrys(mz_zip_internal_state *pState,
```
`s/delete_central_dir_entrys/delete_central_dir_entries/`

jinfeihan57 replied:
OK，I will fix this.

kuba-- commented on src/zip.c:
```diff
@@ -1038,3 +1074,395 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
+}
+
+int zip_entries_delete(struct zip_t *zip, char *entries[], const int num) {
+  if (zip == NULL || entries == NULL) {
```
I think if `entries` is NULL we can do nothing, just return without any error code. The problem is if `entries` is NULL but `num > 0`

jinfeihan57 replied:
OK,  I'll fix it.

kuba-- commented on src/zip.h:
```diff
@@ -336,16 +336,26 @@ extern int zip_extract_stream(const char *stream, size_t size, const char *dir,
+ * @param num the number of entries to be deleted.
+ * @return the number of deleted entries, or negative number (< 0) on error.
+ */
+extern int zip_entries_delete(struct zip_t *zip, char *entries[],
```
Maybe we can replace num by len, so it will be consistent with:
`extern int zip_create(const char *zipname, const char *filenames[], size_t len);`

Also I think it can be `const char *`

jinfeihan57 replied:
ok，I will fix it.

kuba-- reviewed (commented):
Generally LGTM, I left a few comments.

jinfeihan57 commented on src/zip.c:
```diff
@@ -1038,3 +1074,395 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
+    return MZ_FALSE;
+  }
+
+  mz_bool res = MZ_FALSE;
+  if (strcmp(file_name, delete_entry_name) == 0) {
+    res = MZ_TRUE;
+    goto cleanup;
+  }
+
+cleanup:
+  CLEANUP(delete_entry_name);
+  return res;
```
Okay, I'll fix this.

jinfeihan57 commented on src/zip.c:
```diff
@@ -1038,3 +1074,395 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
+
+  mz_zip_archive_file_stat file_stat;
+  mz_uint64 d_pos = ~0;
+  for (int i = 0; i < n; ++i) {
```
"so if we zip_entry_openbyindex, maybe we can (for consistency) just call zip_entry_close after mz_zip_reader_file_stat (instead of CLEANUP here) - does it make sense?" The problem is each time zip_entry_openbyindex() is called, memory allocation for zip->entry.name is generated. If zip_entry_openbyindex() is invoked for multiple times, memory leakage occurs. Before assigning a value to zip->entry.name, check whether the value is NULL. If not, then CLEANUP(zip->entry.name). I'll show it in my next submission.
”I think, if we go from the last to the first one for (int i = n-1; i >= 0; i--) then we can mark DELETE and MOVE and offset in one loop (because we may always keep currently the max offset).
So, in theory we will not need the last loop:” Cause [4.4.1.3 The entries in the central directory MAY NOT necessarily
be in the same order that files appear in the .ZIP file.](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT)，so In the first loop, we need to find the deleted file with the smallest m_local_header_ofs. In the second loop, we need to set the KEEP after the smallest  m_local_header_ofs to MOVE.

jinfeihan57 commented:
@kuba-- Please check again.

kuba-- commented on src/zip.c:
```diff
@@ -1038,3 +1080,394 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
+      entry_mark_array[i].type = KEEP;
+    }
+
+    // CLEANUP(zip->entry.name);
```
leftover

jinfeihan57 replied:
I will delete it.

kuba-- commented on src/zip.c:
```diff
@@ -262,6 +285,9 @@ int zip_entry_open(struct zip_t *zip, const char *entryname) {
     and UNIX file systems etc.  If input came from standard
     input, there is no file name field.
   */
+  if (zip->entry.name) {
+    CLEANUP(zip->entry.name);
+  }
```
Do we need it, or is it _just in case_?

jinfeihan57 replied:
Just in  case.

kuba-- reviewed (commented):
LGTM

jinfeihan57 commented:
@kuba--  It's all fixed.
